### PR TITLE
Add contrastive loss for link prediction tasks.

### PR DESCRIPTION
**Pull request by classicsong**
Add a contrastive loss for link prediction.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by GentleZhu — APPROVED** [review]
LGTM

File: docs/source/configuration/configuration-run.rst [hunk 1]
```diff
@@ -477,10 +477,10 @@ Link Prediction Task
     - Yaml: ``gamma: 10.0``
     - Argument: ``--gamma 10.0``
     - Default value: ``12.0``
-- **lp_loss_func**: Link prediction loss function. Builtin loss functions include ``cross_entropy`` and ``logsigmoid``.
```
Comment (zheng-da) [inline docs/source/configuration/configuration-run.rst:480]: logsigmoid is gone? is this API change?

File: python/graphstorm/config/argument.py [hunk 2]
```diff
@@ -1771,6 +1773,23 @@ def lp_decoder_type(self):
         # Set default value to distmult
         return BUILTIN_LP_DISTMULT_DECODER
 
+    @property
+    def contrastive_loss_temp(self):
```
Comment (zheng-da) [inline python/graphstorm/config/argument.py:1777]: how about using the full name? it's hard for me to think temp is temperature.

File: python/graphstorm/gsf.py [hunk 3]
```diff
@@ -458,44 +463,50 @@ def create_builtin_lp_model(g, config, train_task):
...
+                    if model.gnn_encoder is not None \
+                    else model.node_input_encoder.out_dims
     if config.lp_decoder_type == BUILTIN_LP_DOT_DECODER:
         # if the training set only contains one edge type or it is specified in the arguments,
         # we use dot product as the score function.
         if get_rank() == 0:
             logging.debug('use dot product for single-etype task.')
             logging.debug("Using inner product objective for supervision")
         if config.lp_edge_weight_for_loss is None:
-            decoder = LinkPredictDotDecoder(model.gnn_encoder.out_dims \
```
Comment (zheng-da) [inline python/graphstorm/gsf.py:468]: are we still going to use `LinkPredictDotDecoder` anywhere?

File: python/graphstorm/model/edge_decoder.py [hunk 4]
```diff
@@ -580,9 +582,8 @@ def forward(self, g, h, e_h=None):
                 src_emb = h[src_type][u]
                 dest_emb = h[dest_type][v]
                 scores_etype = calc_dot_pos_score(src_emb, dest_emb)
-                scores.append(scores_etype)
+                scores[canonical_etype] = scores_etype
 
-            scores=th.cat(scores)
```
Comment (zheng-da) [inline python/graphstorm/model/edge_decoder.py:585]: here we actually changed API. is this API exposed to users?

File: python/graphstorm/model/edge_decoder.py [hunk 5]
```diff
@@ -713,6 +714,45 @@ def out_dims(self):
...
+                #
+                # For example:
+                #
+                # pos pairs   |  neg pairs
+                # (10, 20)    |  (10, 3), (10, 1), (10, 0), (10, 22)
+                # (13, 6)     |  (13, 3), (13, 1), (13, 0), (13, 22)
+                # (29, 8)     |  (29, 3), (29, 1), (29, 0), (29, 22)
+                # TODO: use stable to keep the order of negatives. This may not
+                # be necessary
+                u_sort_idx = th.argsort(u, stable=True)
```
Comment (zheng-da) [inline python/graphstorm/model/edge_decoder.py:746]: this is really a temporary solution. the comment above actually doesn't make too much sense to be here because there are no positive and negative pairs in this function. the comment should be in the GNN module where we can see positive edges and negative edges.

File: python/graphstorm/model/edge_decoder.py [hunk 6]
```diff
@@ -713,6 +714,45 @@ def out_dims(self):
         """
         return 1
 
+class LinkPredictContrastiveDotDecoder(LinkPredictDotDecoder):
```
Comment (zheng-da) [inline python/graphstorm/model/edge_decoder.py:717]: once we get to a better solution, this class is no longer needed?

Comment (zheng-da) [reply python/graphstorm/model/edge_decoder.py:717]: if so, we shouldn't expose this API to users. in the docstring, we should say that this is a temporary solution.

File: python/graphstorm/model/loss_func.py [hunk 7]
```diff
@@ -110,9 +110,25 @@ class LinkPredictLossFunc(GSLayer):
```
Comment (zheng-da) [inline python/graphstorm/model/loss_func.py:110]: this class name and the docstring above doesn't make much sense. is this cross-entry loss? we refer to this as link prediction loss?

File: docs/source/configuration/configuration-run.rst (hunk 1 above)
Comment (classicsong) [reply docs/source/configuration/configuration-run.rst:480]: The is no code of logsigmoid. I think we mistakenly add this in our doc.

File: python/graphstorm/model/edge_decoder.py (hunk 4 above)
Comment (classicsong) [reply python/graphstorm/model/edge_decoder.py:585]: It is not exposed.

File: python/graphstorm/model/loss_func.py [hunk 8]
```diff
@@ -172,3 +207,68 @@ def out_dims(self):
...
+            nscore.append(n_s.reshape(p_s.shape[0], -1))
+        pscore = th.cat(pscore, dim=0)
+        nscore = th.cat(nscore, dim=0)
+
+        pscore = th.div(pscore, self._temp)
+        nscore = th.div(nscore, self._temp)
+        score = th.cat([pscore.unsqueeze(1), nscore], dim=1)
+
+        exp_logits = th.exp(score)
+        log_prob = pscore - th.log(exp_logits.sum(1))
```
Comment (zheng-da) [inline python/graphstorm/model/loss_func.py:252]: can you write the formula of the contrastive loss? it's not very clear to me.

File: python/graphstorm/model/edge_decoder.py (hunk 6 above)
Comment (classicsong) [reply python/graphstorm/model/edge_decoder.py:717]: > once we get to a better solution, this class is no longer needed?

I think so. But it may take time for us to implement a better solution.
We can hide it from our users.

Comment (classicsong) [reply python/graphstorm/model/edge_decoder.py:717]: Add a TODO.

File: python/graphstorm/model/edge_decoder.py (hunk 5 above)
Comment (classicsong) [reply python/graphstorm/model/edge_decoder.py:746]: You are right.
It works only because we know how neg_graph is built.

File: python/graphstorm/model/loss_func.py (hunk 7 above)
Comment (classicsong) [reply python/graphstorm/model/loss_func.py:110]: A better name should be LinkPredictBinaryClassLossFunc

File: python/graphstorm/model/loss_func.py (hunk 8 above)
Comment (classicsong) [reply python/graphstorm/model/loss_func.py:252]: Added in the class description

File: python/graphstorm/model/edge_decoder.py (hunk 5 above)
Comment (zheng-da) [reply python/graphstorm/model/edge_decoder.py:752]: can you add a TODO here?

**Review by zheng-da — APPROVED** [review]

File: python/graphstorm/model/edge_decoder.py (hunk 5 above)
Comment (classicsong) [reply python/graphstorm/model/edge_decoder.py:752]: I added a TODO in L724: 
```
TODO(xiang): develop a better solution to unify the implementation of link prediction decoders.
```

Comment (classicsong) [reply python/graphstorm/model/edge_decoder.py:752]: I added a TODO in L725 (class description).
